src/arch/x86/64/c_traps.c

```diff
@@ -39,84 +39,209 @@ static void NORETURN restore_vmx(void)
     /* Do not support breakpoints in VMs, so just disable all breakpoints */
     loadAllDisabledBreakpointState(cur_thread);
 #endif
+    vcpu_restore_guest_msrs(cur_thread->tcbArch.tcbVCPU);
     if (cur_thread->tcbArch.tcbVCPU->launched) {
         /* attempt to do a vmresume */
         asm volatile(
-            // Set our stack pointer to the top of the tcb so we can efficiently pop
-            "movq %[reg], %%rsp\n"
+            // Arguments are getting stored in general purpose registers that need to be used.
```

Comment:
The added assembly code in vmresume and vmlaunch seems quite close. Could the duplication be removed?

Reply:
I don't see why the duplication couldn't be removed